pkg/registrymirror/registrymirrorsstore.go

```diff
@@ -0,0 +1,325 @@
+package registrymirror
+
+import (
+	"bytes"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/BurntSushi/toml"
+	"github.com/containers/image/v5/pkg/sysregistriesv2"
+	ciTypes "github.com/containers/image/v5/types"
+	"github.com/docker/distribution/reference"
+	configV1 "github.com/openshift/api/config/v1"
+	operatorV1Alpha1 "github.com/openshift/api/operator/v1alpha1"
+	"github.com/openshift/runtime-utils/pkg/registries"
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/maputil"
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/stackrox/rox/pkg/utils"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	// default path to store registries config.
+	defaultRegistriesPath = "/var/cache/stackrox/mirrors/registries.conf"
+
+	// default delay before writing the updated registries config.
+	defaultDelay = time.Millisecond * 300
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+// Store defines an interface for interacting with registry mirrors.
+type Store interface {
+	Cleanup()
+
+	UpsertImageContentSourcePolicy(icsp *operatorV1Alpha1.ImageContentSourcePolicy) error
+	DeleteImageContentSourcePolicy(uid types.UID) error
+
+	UpsertImageDigestMirrorSet(idms *configV1.ImageDigestMirrorSet) error
+	DeleteImageDigestMirrorSet(uid types.UID) error
+
+	UpsertImageTagMirrorSet(itms *configV1.ImageTagMirrorSet) error
+	DeleteImageTagMirrorSet(uid types.UID) error
+
+	// PullSources will return image references in the order they should be attempted when pulling the image.
+	// The returned slice may (or may not) include the source image based on the registries config.
+	// Source image must be fully qualified (include the registry hostname).
+	PullSources(srcImage string) ([]string, error)
+}
+
+// FileStore stores/reads the consolidated registries config from a filesystem.
+type FileStore struct {
+	configPath string
+
+	// holds the mirror sets used for managing registry mirrors.
+	icspRules   map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy
+	idmsRules   map[types.UID]*configV1.ImageDigestMirrorSet
+	itmsRules   map[types.UID]*configV1.ImageTagMirrorSet
+	ruleRWMutex sync.RWMutex
+
+	// timer is used when performing delayed writes to filesystem.
```

Comment:
why would we want delayed writes?

Reply:
Writes on most filesystem are not atomic - if writing 'many' things at the same time may get indeterminate results. Writes are also slow.  When sensor first starts up if it gets 100's for mirroring CRs immediately (unlikely) - it didn't make sense to overwrite the consolidated config over and over again potentially stressing out the drive.